### PR TITLE
Improve fsck handling and output

### LINKED ISSUE #2494: fsck --decrypt re-encrypts all the files even those with no need to touch
### Summary
fsck --decrypt reencrypts all the files even those with no need to touch.
I mentioned that allready in #2459 and was told to create a separate issue.

### Steps To Reproduce
gopass fsck --decrypt 

### Expected behavior
should only re-encrypt files where the key changed or such but should ignore files without the need for re-encryption.

### Environment
- OS: fedora 37
- OS version: 6.0.14-300.fc37.x86_64
- gopass Version: master
- Installation method: go install from master used the latest package before
- BTRFS

### PR DESCRIPTION
**Pull request by niacdoial**
- better use of git
- better use of error messages
- smarter behavior around importing public keys from gopass to the user's gpg keyring

Fixes #2494

### CODE REVIEW

**Review by dominikschulz — COMMENTED** [review]
This looks interesting. Thank you.

Just some initial feedback for now. Expect more tomorrow.

- What is the purpose / context of this PR? Please try to provide a little bit more explanation what you're trying to do and why, e.g. what is "QOL"?
- Why did you modify the Queue funcs to return a context?
- What is the error collector doing?
- if else is not very idiomatic Go. If possible we'd try to handle the error case (the one that's less likely) in an if clause and the then continue outside of the if clause. In most cases here that's possible since the if error case would already return or continue.
- Go does not use variables with underscores inside (e.g. err_isfatal).
- Returning strings ("F", "NF") to indicate different severities is non-idiomatic as well. We might have special error types for these or maybe a bool.

**Comment by niacdoial** [comment]
hi,
thank you for reviewing this!

- "QOL" is "Quality of life", so not really new features or bugfixes, just removing some friction from things.
  Basically, there are two ideas:
  - make more messages happen at the end of the fsck process, instead of having them interrupt the progress bar (which then needs to be re-drawn, taking more console space, and making everything harder to read)
  - make sure a single call to fsck results in a single git commit, which tells which secrets were re-written. useful when manual intervention is done in the keyring's repo
- the ability to pass around contexts in the queue allows to send info from one task to the next (although, that was useful when fsck had a non-nil queue, as the context object had its own error collector). This goes well with tasks effectively having two contexts, one from the queue and one from the add-task-to-queue caller.
- the error collector helps delay error messages, so they could be displayed at once not interrupting the progress bar at each message

As for the non-go-idiomatic stuff, I'll see what I can do.

**Comment by niacdoial** [comment]
made it more idiomatic, but it seems that the error message for a non `--decrypt` fsck run is really misleading, so I'll need to fix that (or more likely its root cause) tomorrow.

File: internal/action/delete.go
Comment (dominikschulz) [inline internal/action/delete.go:110]: A more idiomatic approach would be to handle the returned error first and then the warning, like this:

```
if !errors.Is(err, store.ErrMeaninglessWrite) {
  return ...
}
out.Warningf(...)
```

Both are in the error handling branch of Store.Set anyway and any error but MeaninglessWrite should be returned.

This comment applies to this place and any other instance of this pattern.

File: internal/action/generate.go
Comment (dominikschulz) [inline internal/action/generate.go:359]: Do you really want to return with an error if it's an ErrMeaninglessWrite? I don't think so.

File: internal/backend/storage/fs/store.go
Comment (dominikschulz) [inline internal/backend/storage/fs/store.go:66]: Please remove this old debug statement.

File: internal/backend/storage/fs/store.go
Comment (dominikschulz) [inline internal/backend/storage/fs/store.go:65]: This is at the core of what this PR is about. I think this block deserves a small comment explaining why we're doing this.

File: internal/queue/background.go
Comment (dominikschulz) [inline internal/queue/background.go:87]: These files don't seem to be properly formatted.

File: internal/queue/background.go
Comment (dominikschulz) [inline internal/queue/background.go:91]: This should also have a short explanation.

File: internal/store/err.go
Comment (dominikschulz) [inline internal/store/err.go:26]: For the sake of consistency, please change `cyphertext` to `ciphertext`. I think that's how we usually call it in the project.

File: internal/store/leaf/crypto.go
Comment (dominikschulz) [inline internal/store/leaf/crypto.go:59]: This does change the semantic a bit.
Before we would skip the entire loop if we weren't asked to update keys.

But now only if there are any valid keys for this recipient.
Is that really what we want?

File: internal/store/leaf/crypto.go
Comment (dominikschulz) [inline internal/store/leaf/crypto.go:195]: You should just drop the else clause. If the if clause already returned the error then that case won't get there.

Example:

```
if err != nil {
  return err
}
return im.ImportPublicKey(ctx,pk)
```

**Review by dominikschulz — CHANGES_REQUESTED** [review]

File: internal/action/generate.go
Comment (niacdoial) [reply internal/action/generate.go]: yeah, nil is probably more well-suited. anyway, the rest of the function is not needed is the ciphertext secret is unchanged

File: internal/queue/background.go
Comment (niacdoial) [reply internal/queue/background.go:87]: I believe this is now fixed, but I'll let you be the judge of that

File: internal/store/leaf/crypto.go
Comment (niacdoial) [reply internal/store/leaf/crypto.go:59]: the intended change is the following (in pseudocode):
going from:
`if validKeysExist && !wantKeysUpdate {continue}`
to:
`if validKeysExist && (!wantKeysUpdate || bytes.Equals(pubkeyInStore, pubkeyInGPG) ) {continue}`
and the "A or B" part of this condition is moved into the inside of that if block, which only the initial `if validKeysExist` in if statement of that if block.

now that I think about it, the bytewise comparison might not work if there are multiple public keys with that ID. hm…

Comment (niacdoial) [reply internal/store/leaf/crypto.go:59]: well that last remark was dumb.
I just got caught up on the name `s.crypto.FindRecipients` (which can take more than one key ID) and on `if kl > 0`.
Just to be sure, do you confirm that `kl > 1` shouldn't be possible there?

File: internal/store/leaf/link.go
Comment (dominikschulz) [inline internal/store/leaf/link.go:43]: This ctx value is never used.

File: internal/store/leaf/link.go
Comment (dominikschulz) [inline internal/store/leaf/link.go:40]: Most files are not properly formatted. Please use gofmt or goimports.

File: internal/store/leaf/move.go
Comment (dominikschulz) [inline internal/store/leaf/move.go:129]: This value of ctx is never used.

File: internal/store/leaf/reencrypt.go
Comment (dominikschulz) [inline internal/store/leaf/reencrypt.go:62]: printf args are missing.

File: internal/store/leaf/templates.go
Comment (dominikschulz) [inline internal/store/leaf/templates.go:120]: If we don't want to overwrite because it already has the right content shouldn't we consider to return nil instead?

File: internal/store/leaf/write.go
Comment (dominikschulz) [inline internal/store/leaf/write.go:76]: ctx is not used.

File: pkg/ctxutil/ctxutil.go
Comment (dominikschulz) [inline pkg/ctxutil/ctxutil.go:159]: The `is` helper should exaclty do this. Why did you replace it?

File: pkg/ctxutil/ctxutil.go
Comment (dominikschulz) [inline pkg/ctxutil/ctxutil.go:270]: remove the else and outdent the `return false`, please.

File: pkg/ctxutil/ctxutil.go
Comment (dominikschulz) [inline pkg/ctxutil/ctxutil.go:281]: no underscores in variable names, please.

File: pkg/ctxutil/ctxutil.go
Comment (dominikschulz) [inline pkg/ctxutil/ctxutil.go:277]: I'm not super happy with this approach. It should work, but it's neither elegant nor efficient.

What you're doing is:
- get an existing string from the context (if remains in the content)
- concatenate it to create a new strings
- add that to the context (the other value is still there)

In most cases this shouldn't be a big deal, but I can imagine very big password stores where this might become an issue.

So what about another approach: Create a new commit message type (either a custom type with a strings.Builder/bytes.Buffer) or just a ref to one of those. If we only store the reference in the context we can manipulate the value and don't need to add it again on every single invocation.

File: pkg/ctxutil/ctxutil.go
Comment (dominikschulz) [inline pkg/ctxutil/ctxutil.go:327]: no underscores, please

File: pkg/ctxutil/ctxutil.go
Comment (dominikschulz) [inline pkg/ctxutil/ctxutil.go:345]: My remarks about the commit message builder also apply here.
Instead of concatenating strings we might create (or use an existing) multi error type that is referenced here.

But: do we really need to store this in the context?
If there other options are too cumbersome that's find, but I do consider using the context as a last restort. It's not very elegant and we've been trying to recude the use of the context in the past already.

File: internal/action/edit.go
Comment (dominikschulz) [inline internal/action/edit.go:69]: This is a question without a choice. If we do not want to give the user a choice we shouldn't include a question in the first place.

File: internal/store/leaf/templates.go
Comment (niacdoial) [reply internal/store/leaf/templates.go]: done

File: pkg/ctxutil/ctxutil.go
Comment (niacdoial) [reply pkg/ctxutil/ctxutil.go]: no idea why, but I replaced it back

File: pkg/ctxutil/ctxutil.go
Comment (niacdoial) [reply pkg/ctxutil/ctxutil.go]: that part of the code in ctxutil.go was unused anyway, so I just removed it.
(I originally added it before fsck was moved to `WithQueue(nil)`)

**Review by dominikschulz — DISMISSED** [review]
LGTM. Thanks!

**Comment by dominikschulz** [comment]
@niacdoial Thanks a lot. Could you please have a look at the build errors? Then we should be good to go.

File: internal/store/leaf/fsck.go
Comment (dominikschulz) [inline internal/store/leaf/fsck.go:118]: This branch is also executed if there is nothing to commit because nothing was updated. This shouldn't yield an error.

File: internal/store/leaf/reencrypt.go
Comment (dominikschulz) [inline internal/store/leaf/reencrypt.go:44]: There is no out.Println. Use out.Printf, please.

Comment (niacdoial) [reply internal/store/leaf/reencrypt.go]: Used `out.Print()` instead, it also exists

**Comment by niacdoial** [comment]
Concerning the build errors, I have no idea what is happening. I know it builds on my side (with nix, build script modified from the official one, under the cut in case it is relevant)

However, the CI build gives me empty errors when I check the "actions" tab of my copy of gopass (no stdout or stderr visible, but a return code of 2 for the main step of the build for all 3 platforms)

<hr/>

```nix
{ lib
, stdenv
, makeWrapper
, buildGo118Module
, fetchurl
, installShellFiles
, git
, gnupg
, xclip
, wl-clipboard
, passAlias ? false
}:

buildGo118Module rec {
  pname = "gopass";
  version = "dev";

  nativeBuildInputs = [ installShellFiles makeWrapper ];

  src = ./src;
#  src = fetchurl {
#    url = "file:///home/niac/gopasss/src";
#    url = "/home/niac/gopasss/src";
#    rev = "ad18f385f3b95ad8f42b8e6146a6adaf3f52ab59";
#    sha256 = "sha256-0000000000000000000000000000000000000000000=";
#  };

#  vendorSha256 = "sha256-ESIlmY3SzmfWLqLxf62dOhAsHPOKc2YYSOhzpJCjvCQ=";
  vendorSha256 = null;

  subPackages = [ "." ];

  ldflags = [ "-s" "-w" "-X main.version=${version}" "-X main.commit=000000000000" ];

  wrapperPath = lib.makeBinPath (
    [
      git
      gnupg
      xclip
    ] ++ lib.optional stdenv.isLinux wl-clipboard
  );

  postInstall = ''
    installManPage gopass.1
    installShellCompletion --zsh --name _gopass zsh.completion
    installShellCompletion --bash --name gopass.bash bash.completion
    installShellCompletion --fish --name gopass.fish fish.completion
  '' + lib.optionalString passAlias ''
    ln -s $out/bin/gopass $out/bin/pass
  '';

  postFixup = ''
    wrapProgram $out/bin/gopass \
      --prefix PATH : "${wrapperPath}" \
      --set GOPASS_NO_REMINDER true
  '';

  meta = with lib; {
    description = "The slightly more awesome Standard Unix Password Manager for Teams. Written in Go";
    homepage = "https://www.gopass.pw/";
    license = licenses.mit;
    maintainers = with maintainers; [ rvolosatovs sikmir ];
    changelog = "https://github.com/gopasspw/gopass/raw/v${version}/CHANGELOG.md";

    longDescription = ''
      gopass is a rewrite of the pass password manager in Go with the aim of
      making it cross-platform and adding additional features. Our target
      audience are professional developers and sysadmins (and especially teams
      of those) who are well versed with a command line interface. One explicit
      goal for this project is to make it more approachable to non-technical
      users. We go by the UNIX philosophy and try to do one thing and do it
      well, providing a stellar user experience and a sane, simple interface.
    '';
  };
}

```

**Review by dominikschulz — DISMISSED** [review]
LGTM. Thanks a lot!

File: pkg/ctxutil/ctxutil_test.go
Comment (dominikschulz) [inline pkg/ctxutil/ctxutil_test.go:119]: `ctx2` needs to be declared with `ctx2 := ...`.

File: internal/queue/background.go
Comment (dominikschulz) [inline internal/queue/background.go:66]: These files are still not formatted. Can you run `make fmt`, please?

File: pkg/ctxutil/ctxutil_test.go
Comment (niacdoial) [reply pkg/ctxutil/ctxutil_test.go]: done!
(see comment later on)

File: internal/queue/background.go
Comment (niacdoial) [reply internal/queue/background.go]: ran gofmt again, but this time I made sure it acted on all files relevant to this pull request (and also made sure it was the one from go 1.18)

**Comment by niacdoial** [comment]
turns out, the nix build script only runs some of the tests, so that's a bunch of fixes I needed to add.

And CI now builds on windows and macos, but still not linux. (still exit code 2, no stdout, no stderr)

(also I found that rcs_test.go depends on the user having a configured global git name/email to work)

**Comment by dominikschulz** [comment]
Temporarily disabling 'gci' seems to unblock the linters.

Feel free to add ' gci'  to the list of disabled linters in `.golangci.yml` and fix the remaining issues.
Once that looks good I hope we can merge the PR.

**Comment by niacdoial** [comment]
The CI works!
(I didn't know what was the right place to look for failed CI runs, so now I got to fix all the lint hints)
it also seems that the build step on linux depends on the linter

**Review by dominikschulz — APPROVED** [review]
LGTM. Thanks a lot!

**Pull request merged**
